### ISSUE TITLE
Indentation rule flags closure body

### ISSUE THREAD

**Issue by jsotuyod** [CodeNarc/CodeNarc#345]
Given the code:

```groovy
project.files(project.configurations.scaconfig.files.findAll { File it -> it.name.endsWith '.aar' }
    .collect { File it ->
        MessageDigest sha1 = MessageDigest.getInstance('SHA1')
        String inputFile = 'COMMAND=PREPARE_LIBRARY\n' +
            "FILE_PATH=${it.absolutePath}\n" +
            "FILE_SIZE=${it.length()}\n" +
            "FILE_TIMESTAMP=${it.lastModified()}"
        String hash = new BigInteger(1, sha1.digest(inputFile.bytes)).toString(16)
        cacheDir + hash + File.separator + 'output/jars/classes.jar'
    }).asFileTree
```

Codenarc 1.1 flags the whole `collect` closure as incorrectly indented, requiring it to be taken back 1 level to look like:

```groovy
project.files(project.configurations.scaconfig.files.findAll { File it -> it.name.endsWith '.aar' }
    .collect { File it ->
    MessageDigest sha1 = MessageDigest.getInstance('SHA1')
    String inputFile = 'COMMAND=PREPARE_LIBRARY\n' +
        "FILE_PATH=${it.absolutePath}\n" +
        "FILE_SIZE=${it.length()}\n" +
        "FILE_TIMESTAMP=${it.lastModified()}"
    String hash = new BigInteger(1, sha1.digest(inputFile.bytes)).toString(16)
    cacheDir + hash + File.separator + 'output/jars/classes.jar'
    }).asFileTree
```

This erases any visual indication of where the closure ends, making it actually harder to read.

I therefore believe this is a false positive, or at least something common enough to allow the user to customize it.

**Comment by chrismair** [comment]
Looks like this was already fixed for CodeNarc 1.2, presumably by #310 or #313. I added a new test and verified that it failed with CodeNarc 1.1 code and passed now.

**Comment by chrismair** [comment]
I had added a test case with the specified code (under "Given the code:") in the original issue description for CodeNarc 1.2, and it passed. And I just tried running CodeNarc 1.2 against a project containing a file with exactly those contents, and again, no violation.

So, either there seems to be some environmental difference between you and me, or perhaps there is more surrounding or preceding code? I am running Groovy 2.4. Are you running a different Groovy version? Is the above example code still the same? Is that within a method, or in a script file?

**Comment by jsotuyod** [comment]
@chrismair thanks for the quick response. I'm using Groovy 2.4, and this is within a static method in a trait.

The project is OSS, so you can check it out as needed:

https://github.com/Monits/static-code-analysis-plugin/blob/3e67f9251c473fc361e8dc61c6aeead1b50f608b/src/main/groovy/com/monits/gradle/sca/ClasspathAware.groovy#L111-L118

The rule is currently disabled, you can change the config here:

https://github.com/Monits/static-code-analysis-plugin/blob/3e67f9251c473fc361e8dc61c6aeead1b50f608b/config/codenarc/codenarc.xml#L23-L24

That way you can easily reproduce both issues, the `MissingPropertyException` and the bogus violation report.

Let me know if I can help in any other way!

**Comment by chrismair** [comment]
I pulled out the *MissingPropertyException* bug into #355 and checked in a fix for that.

I was able to reproduce the incorrect *Indentation* violations on your project, running locally. I then included the full source for your `ClasspathAware` class in my `IndentationRuleTest`. I was able to reproduce the same 4 violations when running the CodeNarc 1.2 version of `IndentationRule`, and then verified that it ran clean (no violations) with the most recent code. I think (hope) that my recent fixes for #310 also addressed your issues as well.

**Comment by jsotuyod** [comment]
Great! I'm happy you could reproduce it.

I'm looking forward for the next release.

**Comment by chrismair** [comment]
Available in CodeNarc 1.2.1

**Comment by jsotuyod** [comment]
I confirm the issue is resolved on our project. Thanks again for the support!
